For this assignment, we will explore Twitter Emotion Classification. The goal is the identify the primary emotion expressed in a tweet. Consider the following tweets:
```
Tweet 1: @NationalGallery @ThePoldarkian I have always loved this painting.
Tweet 2: '@tateliverpool #BobandRoberta: I am angry more artists that have a profile are not speaking up #foundationcourses.'
``` 

How would you describe the emotions in `Tweet 1` vs `Tweet 2`? `Tweet 1` expresses enjoyment and happiness, while `Tweet 2` directly expresses anger. For this assignment, we will be working with the SMILE Twitter Emotion Dataset ([Wang et al. 2016](https://ceur-ws.org/Vol-1619/paper3.pdf)). At a high level, our goal is to develop different models (rule-based, machine learning, and deep learning), which can be used to identify the emotion of a tweet. You will be required to clean and preprocess the data, generate features for classification, train various models, and evaluate the models. 


*Submission Details*
Please complete all the tasks in “Assignment 1.ipynb” and upload your submission as a Python notebook on Blackboard with the filename “StudentID_Lastname.ipynb”. Assignment 1 will be due by 11:59 PM GMT Monday February 20th, 2023.  

*Grading Policy*

Assignment 1 is graded and will be worth 25% of your overall grade. This assignment is worth a total of 50 points distributed over the tasks below. 
Please note that this is an individual assignment and you must not work with other students to complete this assessment. Any copying from other students, from student exercises from previous years, and any internet resources will not be tolerated. Plagiarised assignments will receive zero marks and the students who commit this act will be reported. 

Feel free to reach out to the TAs and instructors if you have any questions.

Before you get started, run the cell below to download the dataset into memory and a few relevant libraries.

In [1]:
!wget -O data.csv "https://figshare.com/ndownloader/files/4988956"
!pip install emoji

import nltk
nltk.download('punkt')

--2023-02-06 15:43:23--  https://figshare.com/ndownloader/files/4988956
Resolving figshare.com (figshare.com)... 52.212.226.210, 34.254.210.226
Connecting to figshare.com (figshare.com)|52.212.226.210|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/4988956/smileannotationsfinal.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230206/eu-west-1/s3/aws4_request&X-Amz-Date=20230206T154324Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=5e29bb5f5c050191c0e4965e0a9aa2aca0d94eda8387b5cd354a0b4e2277326a [following]
--2023-02-06 15:43:24--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/4988956/smileannotationsfinal.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230206/eu-west-1/s3/aws4_request&X-Amz-Date=20230206T154324Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=5e29bb5f5c050191c0e4965e0a9aa2aca0d94eda8387b5cd354a0b4e2277

[nltk_data] Downloading package punkt to /Users/jiarongli/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [2]:
!wget -O data.csv "https://figshare.com/ndownloader/files/4988956"
!pip install emoji

import nltk
nltk.download('punkt')

--2023-02-06 15:44:00--  https://figshare.com/ndownloader/files/4988956
Resolving figshare.com (figshare.com)... 52.212.226.210, 34.254.210.226
Connecting to figshare.com (figshare.com)|52.212.226.210|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/4988956/smileannotationsfinal.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230206/eu-west-1/s3/aws4_request&X-Amz-Date=20230206T154400Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=2879392e1fea0eb418f8acbcbbae029f662d4afa5a8e8c54537b049265dfd216 [following]
--2023-02-06 15:44:00--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/4988956/smileannotationsfinal.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230206/eu-west-1/s3/aws4_request&X-Amz-Date=20230206T154400Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=2879392e1fea0eb418f8acbcbbae029f662d4afa5a8e8c54537b049265df

[nltk_data] Downloading package punkt to /Users/jiarongli/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

## Task 1. Data Cleaning, Preprocessing, and splitting [15 points]
The `data` environment contains the SMILE dataset loaded into a pandas dataframe object. Our dataset has three columns: id, tweet, and label. The `tweet` column contains the raw scraped tweet and the `label` column contains the annotated emotion category. Each tweet is labelled with one of the following emotion labels:
- 'nocode', 'not-relevant' 
- 'happy', 'happy|surprise', 'happy|sad'
- 'angry', 'disgust|angry', 'disgust' 
- 'sad', 'sad|disgust', 'sad|disgust|angry' 
- 'surprise'

### Task 1a. Label Consolidation [ 3 points]
As we can see above the annotated categories are complex. Several tweets express complex emotions like (e.g. 'happy|sad') or multiple emotions (e.g. 'sad|disgust|angry'). The first things we need to do is clean up our dataset by removing complex examples and consolidating others so that we have a clean set of emotions to predict. 

For Task 1a., write code which does the following:
1. Drops all rows which have the label "happy|sad", "happy|surprise", 'sad|disgust|angry', and 'sad|angry'.
2. Re-label 'nocode' and 'not-relevant' as 'no-emotion'.
3. Re-label 'disgust|angry' and 'disgust' as 'angry'.
4. Re-label 'sad|disgust' as 'sad'.

Your updated `data' dataframe should have 3,062 rows and 5 label categories (no-emotion, happy, angry, sad, and surprise).


In [1]:
# Your Code HERE 
import pandas as pd

data = pd.read_csv("data.csv", header=None) # read the csv file
data = pd.DataFrame(data) # create a dataframe to store the data

# add name to the columns of data
data.columns = ["id", "tweet", "label"]

# Drops all rows which have the label "happy|sad", "happy|surprise", 'sad|disgust|angry', and 'sad|angry'
data = data[(data.label != "happy|sad") & (data.label != 'happy|surprise') & (data.label != "sad|disgust|angry") & (data.label != "sad|angry")]

# Re-label 'nocode' and 'not-relevant' as 'no-emotion'
data.label[data.label == 'nocode'] = 'no-emotion'
data.label[data.label == 'not-relevant'] = 'no-emotion'

# Re-label 'disgust|angry' and 'disgust' as 'angry'.
data.label[data.label == 'disgust|angry'] = 'angry'
data.label[data.label == 'disgust'] = 'angry'

# Re-label 'sad|disgust' as 'sad'.
data.label[data.label == 'sad|disgust'] = 'sad'

/var/folders/2t/_ww03y3x43s96hqdk9grd8qc0000gn/T/ipykernel_38746/2434469729.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.label[data.label == 'not-relevant'] = 'no-emotion'
/var/folders/2t/_ww03y3x43s96hqdk9grd8qc0000gn/T/ipykernel_38746/2434469729.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.label[data.label == 'disgust|angry'] = 'angry'
/var/folders/2t/_ww03y3x43s96hqdk9grd8qc0000gn/T/ipykernel_38746/2434469729.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/us

### Task 1a Tests 
Run the cell below to evaluate your code. To get full credit for this task, your code must pass all tests. Any alteration of the testing code will automatically result in 0 points. 

In [2]:
# Test 1. Data should have 5 unique labels.
print(f"Unique label test: {len(data['label'].unique()) == 5}")

# Test 2. Data labels must be: angry, happy, no-emotion, sad, and surprise
labels = ["angry", "happy", "no-emotion", "sad", "surprise"]
print(f"Label check: { set(data['label'].unique()).difference(labels) == set() }")

# Test 3. Check example counts per label
print(f"Angry example count: {len(data[data['label']=='angry']) == 70}")
print(f"Happy example count: {len(data[data['label']=='happy']) == 1137}")
print(f"No-Emotion example count: {len(data[data['label']=='no-emotion']) == 1786}")
print(f"Sad example count: {len(data[data['label']=='sad']) == 34}")
print(f"Surprise example count: {len(data[data['label']=='surprise']) == 35}")

Unique label test: True
Label check: True
Angry example count: True
Happy example count: True
No-Emotion example count: True
Sad example count: True
Surprise example count: True


### Task 1b. Tweet Cleaning and Processing [10 points]
Raw tweets are noisy. Consider the example below: 
```
'@tateliverpool #BobandRoberta: I am angry more artists that have a profile are not speaking up #foundationcourses. 😠'
```
The mention @tateliverpool and hashtag #BobandRoberta are extra noise that don't directly help with understanding the emotion of the text. The accompanying emoji can be useful but needs to be decoded to it text form :angry: first. 

For this task you will fill complete the `preprocess_tweet` function below with the following preprocessing steps:
1. Lower case all text
2. De-emoji the text
3. Remove all hashtags, mentions, and urls
4. Remove all non-alphabet characters except the followng punctuations: period, exclamation mark, and question mark

Hints: 
- For step 2 (de-emoji), consider using the python [emoji](https://carpedm20.github.io/emoji/docs/) library. The `emoji.demojize` method will convert all emojis to plain text. The `emoji` library is installed in cell [52].
- Follow the processing steps in order. For example calling nltk's word_tokenize before removing hashtags and mentions will end up creating seperate tokens for @ and # and cause problems.

To get full credit for this task, the Test 1b must pass. Only modify the  cell containing the `preprocess_tweet` function and do not alter the testing code block. 

After you are satisfied with your code, run the tests. code to ensure your function works as expected. This cell will also create a new column called `cleaned_tweet` and apply the `preproces_tweet` function to all the examples in the dataset. 

In [3]:
import emoji 
import re

def preprocess_tweet(tweet: str) -> str:
    """
    Function takes a raw tweet and performs the following processing steps:
    1. Lower case all text
    2. De-emoji the text
    3. Remove all hashtags, mentions, and urls 
    4. Remove all non-alphabet characters except the followng punctuations: period, exclamation mark, and question mark
    """
    # 1. lower case all text
    tweet = tweet.lower()
    
    # 2. de-emoji the text
    tweet = emoji.demojize(tweet)
    
    # 3. Remove all hashtags, mentions, and urls 
    #tweet = tweet.split()
    #tweet = re.sub(r'^https?:\/\/.*[\r\n]*', '', tweet, flags=re.MULTILINE)
    tweet = re.sub(r'http\S+', '', tweet)
    tweet = re.sub("@([a-zA-Z0-9_]{1,50})", '', tweet)
    tweet = re.sub("#([a-zA-Z0-9_]{1,50})", '', tweet)
    
    # 4. Remove all non-alphabet characters except the followng punctuations: 
    # period, exclamation mark, and question mark
    tweet = "".join([char for char in tweet
                      if char.isalpha() or char=="." or char=="!" or char=="?" or char==" "])
 
    # remove space
    tweet = tweet.strip()
    tweet = tweet.split()
    tweet = " ".join(tweet)
    
    return tweet 

test_tweet = "'@tateliverpool #BobandRoberta: I am angry more artists that have a profile are not speaking up! #foundationcourses 😠'"
print(preprocess_tweet(test_tweet))

i am angry more artists that have a profile are not speaking up! angryface


### Task 1b Test
Run the cell below to evaluate your code. To get full credit for this task, your code must pass all tests. Any alteration of the testing code will automatically result in 0 points. 

In [4]:
# Do NOT modify the code below. 
# Create new column with cleaned tweets. We will use this for the subsequent tasks
data["cleaned_tweet"] = data["tweet"].apply(preprocess_tweet)

# Test 1b 
test_tweet = "'@tateliverpool #BobandRoberta: I am angry more artists that have a profile are not speaking up! #foundationcourses 😠'"
clean_tweet = "i am angry more artists that have a profile are not speaking up! angryface"
print(f"Test 1b: {preprocess_tweet(test_tweet) == clean_tweet}")

Test 1b: True


### Task 1c. Generating Evaluation Splits [2 points]
Finally, we need to split our data into a train, validation, and test set. We will split the data using a 60-20-20 split, where 60% of our data is used for training, 20% for validation, and 20% for testing. As the dataset is heaviliy imbalanced, make sure you stratify the dataset to ensure that the label distributions across the three splits are roughly equal. 

Store your splits in the variables `train`, `val`, and `test` respectively. 

Hints:
- Use the [`train_test_split`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) function for this task. You'll have to call it twice to get the validation split. 
- Set the random state so the sampling can be reproduced (we use 2023 for our random state)
- Use the `stratify` parameter to ensure representative label distributions across the splits. 

In [5]:
from sklearn.model_selection import train_test_split

# Your code here
train, test = train_test_split(data, random_state=2023, stratify=data["label"], test_size=.20)
train, val = train_test_split(train, random_state=2023, stratify=train["label"], test_size=.20)

## Task 2: Naive Baseline Using a Rule-based Classifier [10 points]

Now that we have a dataset, let's work on developing some solutions for emotion classification. We'll start with implementing a simple rule-based classifier which will also serve as our naive baseline. Emotive language (e.g. awesome, feel great, super happy) can be a strong signal as to the overall emotion being by the tweet. For each emotion in our label space (happy, surprised, sad, angry) we will generate a set of words and phrases that are often associated with that emotion. At classification time, the classifier will calculate a score based on the overlap between the words in the tweet and the emotive words and phrases for each of the emotions. The emotion label with the highest overlap will be selected as the prediction and if there is no match the "no-emotion" label will be predicted. We can break the implementation of this rules-based classifier into three steps:
1. Emotive language extraction from train examples 
2. Developing a scoring algorithm
3. Building the end-to-end classification flow 

### Task 2a. Emotive Language Extraction [4 points] 
For this task you will generate a set of unigrams and bigrams that will be used to predict each of the labels. Using the training data you will need to extract all the unique unigrams and bigrams associated with each label (excluding no-emotion). Then you should ensure that the extracted terms for each emotion label do not appear in the other lists. In the real world, you would then manually curate the generated lists to ensure that associated words were useful and emotive. For the assignment, you won't be required to further curate the generated lists.

Once you've identified the appropiate terms, save them as lists stored in the following environment variables: `happy_words`, `surprised_words`, `sad_words`,and `angry_words`. To get full credit for this section, ensure all 2a Tests pass. 

Hints
- We suggest you use Python's [set methods](https://realpython.com/python-sets/) for this task.
- NLTK has a function for extracting [ngrams](https://www.nltk.org/api/nltk.util.html?highlight=ngrams#nltk.util.ngrams). This function expects a list of tokens as input and will output tuples which you'll need to reconvert into strings. 

In [6]:
# Your code here
from typing import List
from nltk.util import ngrams

# 1. Extract all terms associated with each label
def extract_words(examples: List[str]) -> List[str]:
    """
    Given a list of tweets, return back the unigrams and bigrams found
    across all the tweets. 
    """
    extracted_words_uni_list = []
    extracted_words_bi_list =[]
    
    #for i in range(len(examples)):
    extracted_words_uni =  ngrams(examples.split(), 1)
    extracted_words_bi =  ngrams(examples.split(), 2)

    for _,i in enumerate(extracted_words_uni):
        i = ','.join(i) # convert tuple to string
        extracted_words_uni_list.append(i)

    for _,i in enumerate(extracted_words_bi):
        i = ','.join(i) # convert tuple to string
        extracted_words_bi_list.append(i)
    
    extracted_words = set(extracted_words_uni_list + extracted_words_bi_list)
    
    return extracted_words

happy_words = train.query("label == 'happy'")["cleaned_tweet"].apply(extract_words) 
sad_words = train.query("label == 'sad'")["cleaned_tweet"].apply(extract_words)
angry_words = train.query("label == 'angry'")["cleaned_tweet"].apply(extract_words)
surprise_words = train.query("label == 'surprise'")["cleaned_tweet"].apply(extract_words)

happy_words = set([w for i in happy_words.values.tolist() for w in i])
sad_words = set([w for i in sad_words.values.tolist() for w in i])
angry_words = set([w for i in angry_words.values.tolist() for w in i])
surprise_words = set([w for i in surprise_words.values.tolist() for w in i])


union1 = sad_words.union(angry_words, surprise_words)
union2 = happy_words.union(surprise_words, angry_words) 
union3 = surprise_words.union(happy_words, sad_words)
union4 = angry_words.union(happy_words, sad_words) 

happy_words = happy_words - union1
sad_words = sad_words - union2
angry_words = angry_words - union3
surprise_words = surprise_words - union4

### Task 2a Tests
Run the cell below to evaluate your code. To get full credit for this task, your code must pass all tests. Any alteration of the testing code will automatically result in 0 points. 

In [7]:
# Check sets are non-empty
print("Checking sets are not empty: ")
print(f"Happy words count: {len(happy_words)}, {len(happy_words) > 0}")
print(f"Sad words count: {len(sad_words)}, {len(sad_words) > 0}")
print(f"Angry words count: {len(angry_words)}, {len(angry_words) > 0}")
print(f"Surprise words count: {len(surprise_words)}, {len(surprise_words) > 0}")

# Checks sets are disjoint 
union1 = sad_words.union(angry_words, surprise_words)
union2 = happy_words.union(surprise_words, angry_words) 
union3 = surprise_words.union(happy_words, sad_words)
union4 = angry_words.union(happy_words, sad_words) 

print("\nChecking sets are all disjoint:")
print(f"Happy words disjoint: {happy_words.isdisjoint(union1)}")
print(f"Sad words disjoint: {sad_words.isdisjoint(union2)}")
print(f"Angry words disjoint: {angry_words.isdisjoint(union3)}")
print(f"Surprise words disjoint: {surprise_words.isdisjoint(union4)}")

Checking sets are not empty: 
Happy words count: 7553, True
Sad words count: 329, True
Angry words count: 619, True
Surprise words count: 250, True

Checking sets are all disjoint:
Happy words disjoint: True
Sad words disjoint: True
Angry words disjoint: True
Surprise words disjoint: True


### Task 2b. Scoring using set overlaps [2 points]

Next we will implement to scoring algorithm. Our score will simply be the count of overlapping terms between tweet text and emotive terms. For this task, finish implementing the code below. To get full credit, ensure Test 2b. is successful. 

In [8]:
sample_words = {'cat', 'hat', 'mat', 'bowling', 'bat'}
sample_tweet1 = "that cat is super cool sitting on the mat" 
sample_tweet2 = "the man in the bowling hat sat on the cat"
sample_tweet3 = "the quick brown fox jumped over the lazy dog"


def score_tweet(sample_tweet, sample_words): 
    score = 0
    
    sample_tweet = extract_words(sample_tweet)
        
    for w in sample_tweet:
        if w in sample_words:
            score += 1   
    
    return score


print(f"Test 1: {score_tweet(sample_tweet1, sample_words) == 2}")
print(f"Test 2: {score_tweet(sample_tweet2, sample_words) == 3}")
print(f"Test 3: {score_tweet(sample_tweet3, sample_words) == 0}")

Test 1: True
Test 2: True
Test 3: True


### 2c. Rule-based classification [4 points] 
Let put together our rules-based classfication system. Fill out the logic in the `simple_clf`. Given a tweet, `simple_clf` will generate the overlap score
for each of emotion labels and return the emotion label with the highest score. If there is no match amongst the emotions, the classifier will return 'no-emotion'.

To get full credit for this section, your average F1 score most be greater than 0.

In [9]:
import numpy as np

def simple_clf(tweet: str) -> str:
    """
    Given a tweet, calculate all the emotion overlap scores.
    Return the emotion label which has the largest score. If
    overlap score is 0, return no-emotion. 
    """

    # Your code here 
    score_list = []
    
    happy_score = score_tweet(tweet, happy_words)
    sad_score = score_tweet(tweet, sad_words)
    angry_score = score_tweet(tweet, angry_words)
    surprise_score = score_tweet(tweet, surprise_words)

    score_list.append(happy_score)
    score_list.append(sad_score)
    score_list.append(angry_score)
    score_list.append(surprise_score)
    
    idx_max = np.argmax(score_list)
    
    if max(score_list) == 0:
        preds = "no-emotion"
        
    else:
        if idx_max == 0:
            preds = "happy"
        elif idx_max == 1:
            preds = "sad"
        elif idx_max == 2:
            preds = "angry"
        elif id_max == 3:
            preds = "surprise"
    
    return preds

After finishing the above section, let's evaluate our how model did.

In [10]:
from sklearn.metrics import classification_report

preds = test["cleaned_tweet"].apply(simple_clf)
print(classification_report(test["label"], preds)) 

              precision    recall  f1-score   support

       angry       0.40      0.14      0.21        14
       happy       0.41      0.96      0.57       228
  no-emotion       0.85      0.15      0.26       357
         sad       0.00      0.00      0.00         7
    surprise       0.00      0.00      0.00         7

    accuracy                           0.45       613
   macro avg       0.33      0.25      0.21       613
weighted avg       0.65      0.45      0.37       613



/opt/homebrew/Caskroom/miniconda/base/envs/research/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/homebrew/Caskroom/miniconda/base/envs/research/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/homebrew/Caskroom/miniconda/base/envs/research/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.

## Task 3. Machine learning w/ grammar augmented features [10 points]

Now that we have a naive baseline, let's build a more sophisticated solution using machine learning. Up to this point, we have only considered the words in the tweet as our primary features. The rules-based approach is a very simple bag-of-words classifier. Can we improve performance if we provide some additional linguistic knowledge?

For Task 3 you will do the following:
- Generate part-of-speech features our tweets
- Train two different machine learning classifiers, one with linguistic features and one without
- Evaluate the trained models on the test set

### Task 3a. Grammar Augmented Feature Generation [3 points]
For this task, we will be generating part-of-speech tags for each token in our tweet. Additionally we'll lemmatize the text as well. We will directly include the POS information by appending the tag to the lemma of word itself. For example:
```
Raw Tweet: I am very angry with the increased prices.
POS Augmented Tweet: I-PRP be-VBP very-RB angry-JJ with-IN the-DT increase-VBN price-NNS .-.
```

Complete the `generate_pos_features` using the Spacy library. Once you have an implementation that works, we'll update the `train` and `test` dataframes with a new column called `tweet_with_pos` which contains the output of the `generate_pos_features` method.

In [11]:
import spacy 
from tqdm.notebook import tqdm
nlp = spacy.load("en_core_web_sm")

def generate_pos_features(tweet: str) -> str:
    """
    Given a tweet, return the lemmatized tweet augmented
    with POS tags.
    E.g.:
    Input: "cats are super cool."
    output: "cat-NNS be-VBP super-RB cool-JJ .-."
    """
    
    tweet = nlp(tweet)
    pos_tweet = ""
    
    for i, tok in enumerate(tweet):
        if i == len(tweet)-1:
            pos_tweet += tok.lemma_ + "-" + tok.tag_
        else:
            pos_tweet += tok.lemma_ + "-" + tok.tag_ + " "

    return pos_tweet

sample_tweet = "I hate action movies"
generate_pos_features(sample_tweet)

'I-PRP hate-VBP action-NN movie-NNS'

In [12]:
# Once you have the code working above run this cell.
train["tweet_with_pos"] = train["cleaned_tweet"].apply(generate_pos_features)
test["tweet_with_pos"] = test["cleaned_tweet"].apply(generate_pos_features)

### Task 3a Tests
Run the cell below to evaluate your code. To get full credit for this task, your code must pass all tests. Any alteration of the testing code will automatically result in 0 points. 

In [13]:
sample_texts = [
    ("i am super angry", "I-PRP be-VBP super-RB angry-JJ"),
    ("That movie was great", "that-DT movie-NN be-VBD great-JJ"),
    ("I hate action movies", "I-PRP hate-VBP action-NN movie-NNS")
]
for i, text in enumerate(sample_texts):
    print(f"Test {i+1}: {generate_pos_features(text[0]) == text[1]}")

Test 1: True
Test 2: True
Test 3: True


### Task 3b. Model Training [5 points]
Next we will train two seperate RandomForest Classifier models. For this task you will generate two sets of input features using the `TfidfVectorizer`. We generate Tfidf statistic on the`cleaned_tweet` and the `tweet_with_pos` columns. 

Once you've generated your features, train two different Random Forest classifiers with the generated features and generate the predictions on the test set for each classifier.

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier

# Your code here

tfidf_cleaned_tweet = TfidfVectorizer()

# Fit the tfidf vectorizer on the cleaned text column
tfidf_cleaned_tweet.fit(train["cleaned_tweet"])

# Generate feature for train
train_X_cleaned_tweet = tfidf_cleaned_tweet.transform(train["cleaned_tweet"])
train_y_cleaned_tweet = train["label"].tolist()                # labels are already numericalized 

# Generate features for test
test_X_cleaned_tweet = tfidf_cleaned_tweet.transform(test["cleaned_tweet"])
test_y_cleaned_tweet = test["label"].tolist()


tfidf_pos_tweet = TfidfVectorizer()

# Fit the tfidf vectorizer on the cleaned text column
tfidf_pos_tweet.fit(train["tweet_with_pos"])

# Generate feature for train
train_X_pos_tweet = tfidf_pos_tweet.transform(train["tweet_with_pos"])
train_y_pos_tweet = train["label"].tolist()                # labels are already numericalized 

# Generate features for test
test_X_pos_tweet = tfidf_pos_tweet.transform(test["tweet_with_pos"])
test_y_pos_tweet= test["label"].tolist()


# two seperate RandomForest Classifier models
rfc1 = RandomForestClassifier() # Initialize Model
rfc2 = RandomForestClassifier() # Initialize Model

rfc1.fit(train_X_cleaned_tweet.toarray(), train_y_cleaned_tweet)   # Fit model to train inputs and labels.
rfc2.fit(train_X_pos_tweet.toarray(), train_y_pos_tweet)   # Fit model to train inputs and labels.

preds_cleaned_tweet = rfc1.predict(test_X_cleaned_tweet.toarray())   # Generate Predictions 
preds_pos_tweet = rfc2.predict(test_X_pos_tweet.toarray())   # Generate Predictions 

### Task 3c. [2 points]
Generate classification reports for both models. Print the reports below. In a few sentences (no more than 100 words) explain which features were the most effective and why you think that's the case?

In [15]:
from sklearn.metrics import classification_report

# Classification Report for Tfidf features
print("Classification report for TFIDF features")
# Your code here
print(classification_report(test_y_cleaned_tweet, preds_cleaned_tweet))  # Generate Classification Report

# Classfication Report for POS features 
print("Classification report for TFIDF w/ POS features")
# Your code here
print(classification_report(test_y_pos_tweet, preds_pos_tweet))  # Generate Classification Report

Classification report for TFIDF features
              precision    recall  f1-score   support

       angry       0.00      0.00      0.00        14
       happy       0.84      0.71      0.77       228
  no-emotion       0.79      0.93      0.85       357
         sad       0.00      0.00      0.00         7
    surprise       0.00      0.00      0.00         7

    accuracy                           0.81       613
   macro avg       0.33      0.33      0.32       613
weighted avg       0.77      0.81      0.78       613

Classification report for TFIDF w/ POS features
              precision    recall  f1-score   support

       angry       0.00      0.00      0.00        14
       happy       0.78      0.70      0.74       228
  no-emotion       0.78      0.90      0.84       357
         sad       0.00      0.00      0.00         7
    surprise       0.00      0.00      0.00         7

    accuracy                           0.78       613
   macro avg       0.31      0.32      0.3

/opt/homebrew/Caskroom/miniconda/base/envs/research/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/homebrew/Caskroom/miniconda/base/envs/research/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/homebrew/Caskroom/miniconda/base/envs/research/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.


From the classification report, feature 'happy' and feature 'no-emotion' are the most effective features. One of the reasons is that the training dataset we use is imbalanced. As we can see from the next cell, the number of instances associated with the label happy and label no-emotion are far more than the number of instances with other labels in the training dataset. In addition, the test dataset is also imbalanced. Therefore, the trained model tends to make more decisions on these two labels.

In [16]:
train.groupby("label").aggregate("count")

id  tweet  cleaned_tweet  tweet_with_pos
label                                                 
angry         45     45             45              45
happy        727    727            727             727
no-emotion  1143   1143           1143            1143
sad           22     22             22              22
surprise      22     22             22              22

In [17]:
test.groupby("label").aggregate("count")

id  tweet  cleaned_tweet  tweet_with_pos
label                                                
angry        14     14             14              14
happy       228    228            228             228
no-emotion  357    357            357             357
sad           7      7              7               7
surprise      7      7              7               7

## Task 4. Transfer Learning with DistilBERT [10 points]

For this task you will finetune a pretrained language model (DistilBERT) using the huggingface `transformers` library. For this task you will need to:
- Encode the tweets using the BERT tokenizer
- Create pytorch datasets for for the train, val and test datasets
- Finetune the distilbert model for 5 epochs
- Extract predictions from the model's output logits and convert them into the emotion labels.
- Generate a classification report on the predictions.

Ensure you are running the notebook in Google Colab with the gpu runtime enabled for this section.

In [18]:
!pip install transformers >> NULL

In [19]:
from sklearn.preprocessing import LabelEncoder
import torch
from torch.utils.data import Dataset
from transformers import AutoTokenizer
from transformers import AutoModelForSequenceClassification
from transformers import Trainer
from transformers import TrainingArguments


# Your Code here

# label encoder
le = LabelEncoder()
le.fit(train["label"])
train["encoded_label"] = le.transform(train["label"])
val["encoded_label"] = le.transform(val["label"])
test["encoded_label"] = le.transform(test["label"])

# convert them to pytorch tensor
train_labels = torch.tensor(train["encoded_label"].tolist())
val_labels = torch.tensor(val["encoded_label"].tolist())
test_labels = torch.tensor(test["encoded_label"].tolist())

# tweet encoding
tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")

train_encodings = tokenizer(
    train["cleaned_tweet"].tolist(),
    padding=True,           # pad all inputs to max length
    max_length=280,         # max size character in tweet
    return_tensors="pt",    # Return format pytorch tensor
    truncation=True
)

val_encodings = tokenizer(
    val["cleaned_tweet"].tolist(),
    padding=True,           # pad all inputs to max length
    max_length=280,         # max size character in tweet
    return_tensors="pt",    # Return format pytorch tensor
    truncation=True
)

test_encodings = tokenizer(
    test["cleaned_tweet"].tolist(),
    padding=True,           # pad all inputs to max length
    max_length=280,         # max size character in tweet
    return_tensors="pt",    # Return format pytorch tensor
    truncation=True
)

# Update encodings with labels
train_encodings["labels"] = train_labels
val_encodings["labels"] = val_labels
test_encodings["labels"] = test_labels


# Create pytorch datasets for for the train, val and test datasets
class TweetDataset(Dataset):
    
    def __init__(self, encodings: dict):  
        self.encodings = encodings
        
    def __len__(self) -> int:
        return len(self.encodings["input_ids"])
    
    def __getitem__(self, idx: int) -> dict:
        e = {k: v[idx] for k,v in self.encodings.items()}
        return e 
    
# Generate Datasets
train_ds = TweetDataset(train_encodings)
val_ds = TweetDataset(val_encodings)
test_ds = TweetDataset(test_encodings)

# Finetune the distilbert model for 5 epochs
model = AutoModelForSequenceClassification.from_pretrained("distilbert-base-uncased", num_labels=7)

training_args = TrainingArguments(
    output_dir="./results",
    num_train_epochs=5,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    lr_scheduler_type='cosine',
    per_device_train_batch_size = 32,
    per_device_eval_batch_size = 32, 
    #fp16=True,
)

trainer = Trainer(
    model,
    training_args,
    train_dataset=train_ds,
    eval_dataset=val_ds,
)

trainer.train()

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForSequenceClassification: ['vocab_transform.weight', 'vocab_layer_norm.bias', 'vocab_transform.bias', 'vocab_layer_norm.weight', 'vocab_projector.bias', 'vocab_projector.weight']
- This IS expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.weight', 'pre_classifier.weight', 'classifi

***** Running Evaluation *****
  Num examples = 490
  Batch size = 32
Saving model checkpoint to ./results/checkpoint-62
Configuration saved in ./results/checkpoint-62/config.json
Model weights saved in ./results/checkpoint-62/pytorch_model.bin
***** Running Evaluation *****
  Num examples = 490
  Batch size = 32
Saving model checkpoint to ./results/checkpoint-124
Configuration saved in ./results/checkpoint-124/config.json
Model weights saved in ./results/checkpoint-124/pytorch_model.bin
***** Running Evaluation *****
  Num examples = 490
  Batch size = 32
Saving model checkpoint to ./results/checkpoint-186
Configuration saved in ./results/checkpoint-186/config.json
Model weights saved in ./results/checkpoint-186/pytorch_model.bin
***** Running Evaluation *****
  Num examples = 490
  Batch size = 32
Saving model checkpoint to ./results/checkpoint-248
Configuration saved in ./results/checkpoint-248/config.json
Model weights saved in ./results/checkpoint-248/pytorch_model.bin
***** Runni

TrainOutput(global_step=310, training_loss=0.32519715832125756, metrics={'train_runtime': 542.1981, 'train_samples_per_second': 18.065, 'train_steps_per_second': 0.572, 'total_flos': 217961938565460.0, 'train_loss': 0.32519715832125756, 'epoch': 5.0})

In [20]:
# Extract predictions from the model's output logits and convert them into the emotion labels.
preds = trainer.predict(test_ds)
preds = le.inverse_transform(np.argmax(preds.predictions, axis=1))

# Generate a classification report on the predictions.
print(classification_report(test["label"].tolist(), preds))

***** Running Prediction *****
  Num examples = 613
  Batch size = 32


              precision    recall  f1-score   support

       angry       0.60      0.43      0.50        14
       happy       0.86      0.90      0.88       228
  no-emotion       0.90      0.91      0.90       357
         sad       1.00      0.14      0.25         7
    surprise       0.00      0.00      0.00         7

    accuracy                           0.88       613
   macro avg       0.67      0.48      0.51       613
weighted avg       0.87      0.88      0.87       613



## Task 5. Model Recommendation [5 points]
In a paragraph (no more than 250 words) answer the following questions:
1. Which of the implemented models would you recommend and why? 
2. Compare the metrics for each models implemted (Rules-Based, Machine Learning w/ POS features, and DistilBERT). What are the pros and con for each model (consider performance both macro performance and label specifc metrics and the computational requirements). 

In this case, I would recommend the pre-trained DistilBERT model instead of the RandomForestClassification model and Rule-Based model. Because it has better performance than the results of RandomForestClassification as we can see from the classification report. In addition, DistilBERT is powerful as it retains 90% of BERT's performance. The Rule-Based model implemented in this case has the lowest accuracy. The advantage of this model is fast decision-making, however, the number of classes it has is limited and it won't learn when there is a new class appears. The RandomForestClassification model we implemented can learn the knowledge during the training process but is limited in the imbalanced datasets. The DistilBERT model we implement can learn knowledge during the training process and it can decrease the influence of the imbalanced dataset as it consists of a pre-trained model and a classifier. Therefore, it can just fine-tune the parameters to have good performance. However, DistilBERT requires a longer time to train and its results are not very accurate as it is a general model instead of a specific domain model.